File: Trees/ConstructBinaryTreeFromInorderAndPostorderTraversal.go
// LeetCode 106 - Construct Binary Tree from Inorder and Postorder Traversal
// Time Complexity: O(n) | Space Complexity: O(n)
package main

func buildTree(inorder []int, postorder []int) *TreeNode {
	inorderIndex := make(map[int]int, len(inorder))

	for index, value := range inorder {
		inorderIndex[value] = index
	}

	var dfs func(int, int, int) *TreeNode
	dfs = func(inorderStart int, postorderStart int, size int) *TreeNode {
		if size <= 0 {
			return nil
		}

		rootValue := postorder[postorderStart+size-1]
		splitIndex := inorderIndex[rootValue]
		left := dfs(inorderStart, postorderStart, splitIndex-inorderStart)
		right := dfs(
			splitIndex+1,
			postorderStart+splitIndex-inorderStart,
			size-splitIndex+inorderStart-1,
		)

		return &TreeNode{Val: rootValue, Left: left, Right: right}
	}

	return dfs(0, 0, len(inorder))
}

File: Trees/ConstructBinaryTreeFromInorderAndPostorderTraversal.py
# LeetCode 106 - Construct Binary Tree from Inorder and Postorder Traversal
# Time Complexity: O(n) | Space Complexity: O(n)
from typing import List, Optional


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        inorder_index = {value: index for index, value in enumerate(inorder)}

        def dfs(inorder_start: int, postorder_start: int, size: int) -> Optional[TreeNode]:
            if size <= 0:
                return None

            root_value = postorder[postorder_start + size - 1]
            split_index = inorder_index[root_value]
            left = dfs(inorder_start, postorder_start, split_index - inorder_start)
            right = dfs(
                split_index + 1,
                postorder_start + split_index - inorder_start,
                size - split_index + inorder_start - 1,
            )
            return TreeNode(root_value, left, right)

        return dfs(0, 0, len(inorder))
